### DISCUSSION TITLE
immerAtom with useReducerAtom type not match

### DISCUSSION THREAD

**Discussion by linkjane** [pmndrs/jotai#956] [Q&A]
jotai is amazing, but I have met some problem; 

![image](https://user-images.githubusercontent.com/15264018/149089655-f6a63e6f-b8c1-4460-869e-a7c096e32373.png)

immerAtom with useReducerAtom type not match ,  But it's runtime works

and another problem is Provider with initalValue when use react fast update (hmr) the initalValue will not apply at not

![image](https://user-images.githubusercontent.com/15264018/149090780-55e65d70-e2aa-4323-88da-d829be024589.png)

thank you for helping

**Comment by dai-shi** [comment]
Hi, thanks for trying out jotai!

> immerAtom with useReducerAtom type not match , But it's runtime works

I didn't know even it works. It's never intended.
Why do you not just use `useAtom`?

**Reply by linkjane** [reply]
Because I want use it like useImmerReducer,  and the library of jotai/immer doesn't has useImmerReduer.

**Reply by dai-shi** [reply]
I see. That's not covered. But you can easily create such a hook on your end.

```js
const useImmerReducer = (anAtom, reducer) => {
  const [state, setStateWithImmer] = useImmerAtom(anAtom)
  const dispatch = useCallback((action) => {
    setStateWithImmer((draft) => reducer(draft, action))
  }, [setStateWithImmer, reducer])
  return [state, dispatch]
}
```

**Reply by linkjane** [reply]
Great!  it works, and I will do some research on the correct type definition.
thank you so much!

**Comment by dai-shi** [comment]
> when use react fast update (hmr)

HMR is quite tough. Please define the atom in a separate file, and if you don't modify that file, it may work. Alternatively, babel plugin https://jotai.org/docs/api/babel#plugin-react-refresh is something to try to solve the issue.

**Reply by linkjane** [reply]
Ok, thank you so much, and I will try to do it. in other words, jotai is awsome!
